sdk/core/azure_core/src/task/mod.rs

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+//! Asynchronous task execution utilities.
+
+use std::{fmt, future::Future, pin::Pin, sync::Arc};
+
+#[cfg(all(not(feature = "tokio"), not(target_arch = "wasm32")))]
+mod standard_spawn;
+
+#[cfg(all(not(feature = "tokio"), target_arch = "wasm32"))]
+mod wasm_spawn;
+
+#[cfg(feature = "tokio")]
+mod tokio_spawn;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(all(not(feature = "tokio"), not(target_arch = "wasm32")))]
+pub use standard_spawn::StdSpawner;
+
+#[cfg(all(not(feature = "tokio"), target_arch = "wasm32"))]
+pub use wasm_spawn::WasmSpawner;
+
+#[cfg(feature = "tokio")]
+pub use tokio_spawn::TokioSpawner;
+
+#[cfg(not(target_arch = "wasm32"))]
+pub(crate) type TaskFuture = Pin<Box<dyn Future<Output = ()> + Send + 'static>>;
+
+#[cfg(target_arch = "wasm32")]
+pub(crate) type TaskFuture = Pin<Box<dyn Future<Output = ()> + 'static>>;
+
+/// Creates a new [`TaskSpawner`].
+///
+/// This returns a [`TaskSpawner`] that spawns a [`std::thread`] to run the task unless `tokio` was enabled,
+/// in which case an executor is returned that calls [`tokio::task::spawn`].
+pub fn new_task_spawner() -> Arc<dyn TaskSpawner> {
+    #[cfg(feature = "tokio")]
+    {
+        Arc::new(TokioSpawner)
+    }
+    #[cfg(all(not(feature = "tokio"), not(target_arch = "wasm32")))]
+    {
+        Arc::new(StdSpawner)
+    }
+    #[cfg(all(not(feature = "tokio"), target_arch = "wasm32"))]
+    {
+        Arc::new(WasmSpawner)
+    }
+}
+
+#[cfg(feature = "tokio")]
+#[derive(Debug)]
+pub struct SpawnHandle(tokio::task::JoinHandle<()>);
+
+#[cfg(feature = "tokio")]
+impl SpawnHandle {
+    /// Wait for the task to complete and return the result.
+    pub async fn await_result(self) -> crate::Result<()> {
+        self.0.await.map_err(|e| {
+            crate::Error::message(
+                crate::error::ErrorKind::Other,
+                format!("Task was cancelled before completion: {}", e),
+            )
+        })
+    }
+}
+
+#[cfg(all(not(feature = "tokio"), not(target_arch = "wasm32")))]
```

Comment:
Rather than repeat this condition over and over, either put it in the `tokio_spawn` module and export it here with that condition, or make a private module and export everything from there. Much easier to manage and read:

```rust
#[cfg(...)]
mod __tokio {
    pub struct ...;
}

#[cfg(same ...)]
pub use __tokio::*;
```